text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1018 @@
+
+The approach for OTLP traces is similar to that used for logs. The primary `SPANS` entity (i.e., Arrow record)
+encompasses a flattened representation of `ResourceSpans`, `ScopeSpans`, and `Spans`. Beyond the standard set of
+attributes (i.e., resource, scope, and span attributes), this mapping represents span events and span links as distinct
+entities (`SPAN_EVENTS` and `SPAN_LINKS` respectively). These have a 1-to-many relationship with the `SPANS` entity.
+Each of these entities is also associated with dedicated attribute entities (i.e. `SPAN_EVENT_ATTRS` and
+`SPAN_LINK_ATTRS`).
+
+![Traces Arrow Schema](img/0156_traces_schema.png)
+
+Similarly, each of the Arrow records is sorted by specific columns to optimize the compression ratio.
+
+The `end_time_unix_nano` is represented as a duration (`end_time_unix_nano` - `start_time_unix_nano`) to reduce the
+number of bits required to represent the timestamp.
+
+#### Metrics Arrow Mapping
+
+The mapping for metrics, while being the most complex, fundamentally follows the same logic as applied to logs and
+spans. The primary 'METRICS' entity encapsulates a flattened representation of `ResourceMetrics`, `ScopeMetrics`, and
+`Metrics`. All common columns among the different metric types are consolidated in this main entity (i.e., `metric_type`,
+`name`, `description`, `unit`, `aggregation_temporality`, and `is_monotonic`). Furthermore, a dedicated entity is
+crafted to represent data points for each type of metrics, with their columns being specific to the respective metric
+type. For instance, the `SUMMARY_DATA_POINTS` entity includes columns `id`, `parent_id`, `start_time_unix_nano`,
+`time_unix_nano`, `count`, `sum`, and `flags`. Each of these "data points" entities is linked to:
+
+- A set of data point attributes (following a one-to-many relationship).
+- A set of data points exemplars (also adhering to a one-to-many relationship).
+
+Exemplar entities, in turn, are connected to their dedicated set of attributes.
+
+Technically speaking, the `quantile` entity isn't encoded as an independent entity but rather as a list of struct within
+the `SUMMARY_DATA_POINTS entity`.
+
+![Metrics Arrow Schema](img/0156_metrics_schema.png)
+
+Gauge and Sum are identified by the `metric_type` column in the `METRICS` entity and they share the same Arrow record
+for the data points, i.e. `NUMBER_DATA_POINTS`.
+
+`span_id` and `trace_id` are represented as fixed size binary dictionaries by default but can evolve to non-dictionary
+form when their cardinality exceeds a certain threshold (usually 2^16).
+
+As usual, each of these Arrow records is sorted by specific columns to optimize the compression ratio. With this mapping
+batch of metrics containing a large number of data points sharing the same attributes and timestamp will be highly
+compressible (multivariate time-series scenario).
+
+> Note: every OTLP timestamps are represented as Arrow timestamps with nanoseconds time unit. This representation will
```

Comment:
```suggestion
> Note: every OTLP timestamps are represented as Arrow timestamps with nanoseconds time unit. This representation will
```
are those nanoseconds since epoch? No time zone encoding so utc only, correct?

Reply:
By default Arrow Timestamp are timezone "naive" so UTC only.